js/web/lib/wasm/wasm-training-core-impl.ts

```diff
@@ -0,0 +1,162 @@
+      const names = [];
+      const wasm = getInstance();
+
+      const namesUTF8Encoded = [];
+
+      for (let i = 0; i < count; i++) {
+        if (wasm._OrtTrainingGetModelInputOutputName) {
+          const name = wasm._OrtTrainingGetModelInputOutputName(trainingSessionId, i, isInput, isEvalModel);
+          if (name === 0) {
+            checkLastError('Can\'t get input or output name');
+          }
+
+          namesUTF8Encoded.push(name);
+          names.push(wasm.UTF8ToString(name));
+        } else {
+          throw new Error(NO_TRAIN_FUNCS_MSG);
+        }
+      }
+      return [names, namesUTF8Encoded];
+    };
+
+const getTrainingModelInputOutputNames = (trainingSessionId: number): [string[], number[], string[], number[]] => {
+  const [inputCount, outputCount] = getModelInputOutputCount(trainingSessionId, false);
+
+  const [inputNames, inputNamesUTF8Encoded] = getModelInputOutputNamesLoop(trainingSessionId, inputCount, true, false);
+  const [outputNames, outputNamesUTF8Encoded] =
+      getModelInputOutputNamesLoop(trainingSessionId, outputCount, false, false);
+
+  return [inputNames, inputNamesUTF8Encoded, outputNames, outputNamesUTF8Encoded];
+};
+
+export const createTrainingSessionHandle =
+    (checkpointHandle: number, trainModelData: SerializableModeldata, evalModelData: SerializableModeldata,
+     optimizerModelData: SerializableModeldata,
+     options: InferenceSession.SessionOptions): [SerializableSessionMetadata, number[], number[]] => {
+      const wasm = getInstance();
+
+      let trainingSessionHandle = 0;
+      let sessionOptionsHandle = 0;
+      let allocs: number[] = [];
+      let inputNamesUTF8Encoded: number[] = [];
+      let outputNamesUTF8Encoded: number[] = [];
+
+      let inputNames: string[] = [];
+      let outputNames: string[] = [];
+
+      try {
+        [sessionOptionsHandle, allocs] = setSessionOptions(options);
+        if (wasm._OrtTrainingCreateSession) {
+          trainingSessionHandle = wasm._OrtTrainingCreateSession(
+              sessionOptionsHandle, checkpointHandle, trainModelData[0], trainModelData[1], evalModelData[0],
+              evalModelData[1], optimizerModelData[0], optimizerModelData[1]);
+        } else {
+          throw new Error(NO_TRAIN_FUNCS_MSG);
+        }
+
+        if (trainingSessionHandle === 0) {
+          checkLastError('Error occurred when trying to create a TrainingSession.');
+        }
+
+        [inputNames, inputNamesUTF8Encoded, outputNames, outputNamesUTF8Encoded] =
+            getTrainingModelInputOutputNames(trainingSessionHandle);
+        return [[trainingSessionHandle, inputNames, outputNames], inputNamesUTF8Encoded, outputNamesUTF8Encoded];
+
+      } catch (e) {
+        if (wasm._OrtTrainingReleaseSession && trainingSessionHandle !== 0) {
+          wasm._OrtTrainingReleaseSession(trainingSessionHandle);
+        }
+        throw e;
+      } finally {
+        wasm._free(trainModelData[0]);
+        wasm._free(evalModelData[0]);
```

Comment:
what happens when eval model or optimizer model is not provided?

Reply:
the TrainingSessionHandler passes 2 nullpointers when evalModel or optimizerModel isn't provided, so the nullpointer would be freed